keps/sig-apps/20190410-configmap-secret-orchestration.md

```diff
@@ -0,0 +1,214 @@
+
+We would like to solve these issues by adding support to the orchestration resources to support orchestration support around config/secrets.
+
+
+## Motivation
+
+Currently there are a mishmash of workarounds to the orchestration objects not supporting orchestration of configmaps/secrets.
+
+Several different solutions or workarounds out of tree are being used. See Alternatives section for details. None of these are widely deployed. Since most workloads need this kind of orchestration, having it as a base Kubernetes functionality would be a significant benefit to users.
+
+### Goals
+
+* Enhance the deployment/statefulset/daemonset with a flag asking for the workload to be automatically updated when a specified configmap/secret changes.
+* Enhance the deployment/statefulset/daemonset with a flag asking for the specified configmap/secret's current state to be snapshotted and follow the orchestrated workload (roll-forward/roll-back)
+* Implementations of the api changes
+
+### Non-Goals
+
+* Out of tree solutions. For applications to be portable, it really needs to be in Kubernetes proper so application developers know the functionality is always there.
+
+## Proposal
+
+The ConfigMapVolumeSource and SecretVolumeSource objects will get two new, optional fields:
+ * snapshot: boolean
+ * watch: boolean
+
+Both will default to false for compatibility.
+
+snapshot can be either true or false. Watch can only be true if snapshot is also true. This ensures immutability.
+
+The fields will be ignored by all objects other then Deployment, DaemonSet and StatefulSet.
+
+DaemonSet/StatefulSet controllers will be modified such that:
+ * During a "change", on seeing a snapshot=true flag on a configmap/secret, a copy of the configmap/secret will be made with a unique name. This unique name will be stored in the corresponding ControllerRevision.
+ * All pods created will reference the unique configmap/secret snapshot name, not the base name.
+ * When a ControllerRevision is deleted, a deletion of the corresponding configmap/secret snapshots will also be issued.
+ * When an object with a watch flag of true is created/updated, watches on the specified configmap/secret will be added. Any watch triggered will be treated as a "change" of the object.
+
+The Deployment controller will be modified such that:
+ * During a "change", on seeing a snapshot=true flag on a configmap/secret, a copy of the configmap/secret will be made with a unique name. This unique name will be stored in the corresponding ReplicaSet.
+ * When a ReplicaSet is deleted, a deletion of the corresponding configmap/secret snapshots will also be issued.
```

Comment:
How could this work for Jobs, or for Pods which are not controlled by a workload controller?

Reply:
It could be made to work with a Job.

Pods though it shouldn't. The issue really is, from the time a workload request is created, ReplicaSet, DaemonSetVersion, Job, etc to the time its pods start, there is a window where configmaps/secrets can be changed. This causes skew in the workload. This snapshot feature keeps the pods managed by those objects always consistent with the desired state at the time of workload creation, not at pod creation. As pods can randomly be recreated across the cluster at any time.

Reply:
It can not be made to work on pods, pod volumes are immutable. I would also argue it should not be made on pods, because that doesn't allow to have any control in case the new revision causes issue.

I think more or less the same is true for Jobs, you should use immutable configmaps/secrets for them to get deterministic results

Reply:
Hmm... Yeah. for jobs, snapshotting the configmap/secret makes sense. Watching, I don't think so? A watch would mostly cause it to revert to existing behavior?